### PR TITLE
Test Revise on buildkite

### PR DESCRIPTION
**Pull request by KristofferC**
It's been discussed that Revise should also be tested on CI because the effect of the productivity of Julia itself significantly goes down when Revise is not working.

Also:

image

Someone feel free to take over this PR.

### CODE REVIEW

**Comment by KristofferC** [comment]
For the Revise test:

```
Precompiling all packages...
--
  | 394.6 ms  ✓ Unicode
  | 496.0 ms  ✓ Requires
  | 529.2 ms  ✓ OrderedCollections
  | 897.6 ms  ✓ Serialization
  | 1641.0 ms  ✓ InteractiveUtils
  | 1768.2 ms  ✓ Distributed
  | 1143.9 ms  ✓ CodeTracking
  | 8662.0 ms  ✓ JuliaInterpreter
  | 6836.8 ms  ✓ LoweredCodeUtils
  | 31739.1 ms  ✓ REPL
  | 7050.6 ms  ✓ Revise
...


Precompiling all packages for configuration --code-coverage=none --color=yes --check-bounds=yes --warn-overwrite=yes --depwarn=yes --inline=yes --startup-file=no --track-allocation=none...
--

  | 38008.7 ms  ✓ REPL
  | 7243.2 ms  ✓ Revise
  | 60380.3 ms  ✓ Pkg
  | 4501.5 ms  ✓ Pkg → REPLExt
```

I wonder why the stdlib precompile files are not valid.

File: pipelines/main/misc/test_revise.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+                - "/cache/repos:/cache/repos"
+        commands: |
+          # Download pre-built julia, extract into `usr/`
+          buildkite-agent artifact download --step "build_x86_64-linux-gnu" 'julia-*-linux-x86_64.tar.gz' .
+          mkdir -p ./usr
+          tar -C ./usr --strip-components=1 -zxf julia-*-linux-x86_64.tar.gz
+          ln -s ./usr/bin/julia ./julia
+
+          echo "--- Install and test Revise"
+          ./julia -e 'using Pkg; Pkg.add("Revise"); Pkg.test("Revise")'
```
Comment (IanButterworth) [inline pipelines/main/misc/test_revise.yml:32]: Print `DEPOT_PATH` here to check? and try with `JULIA_DEBUG=loading`?

Comment (IanButterworth) [reply pipelines/main/misc/test_revise.yml]: bingo 
```
DEPOT_PATH = ["/cache/julia-buildkite-plugin/depots/8c032ad4-259f-4c1c-9f4f-7d5a0ea47bcd"]
```

Comment (IanButterworth) [reply pipelines/main/misc/test_revise.yml]: Add `unset JULIA_DEPOT_PATH`?
as per https://github.com/JuliaCI/julia-buildkite/blob/747ec36f50e816871364680cbd442b4e14632a2a/utilities/test_julia.sh#L48

**Comment by DilumAluthge** [comment]
I haven't reviewed the PR diff, but I agree with the goal of testing Revise itself as part of the Buildkite CI.

**Comment by DilumAluthge** [comment]
@IanButterworth Is it necessary to re-sign the pipelines? AFAICT this PR only touches unsigned pipelines?

**Comment by IanButterworth** [comment]
Ah ok. Feel free to revert that commit

**Comment by KristofferC** [comment]
```
readdir((Revise.git_repo(Revise.juliadir))[2]) = [".buildkite", ".buildkite-external-version", ".clang-format", ".clangd", ".codecov.yml", ".devcontainer", ".git", ".git-blame-ignore-revs", ".gitattributes", ".github", ".gitignore", ".mailmap", "CITATION.bib", "CITATION.cff", "CONTRIBUTING.md", "HISTORY.md", "LICENSE.md", "Make.inc", "Makefile", "NEWS.md", "README.md", "THIRDPARTY.md", "VERSION", "base", "cli", "contrib", "deps", "doc", "etc", "julia", "julia-df1976d500-linux-x86_64.tar.gz", "julia.spdx.json", "pkgimage.mk", "src", "stdlib", "sysimage.mk", "test", "typos.toml", "usr"]
```

Where is this julia git repo that we are standing in coming from?

**Comment by staticfloat** [comment]
> Where is this julia git repo that we are standing in coming from?

Because this is a pull request against the `julia-buildkite` repo, not the `julia` repo, buildkite automatically clones `julia-buildkite`, not `julia` [0].  So we add a local `post-checkout` hook to clear out the working directory and clone Julia [1], then re-clone the `julia-buildkite` repository (because that's part of the CI scripts here) [2] and make it look as much as possible like what the base Julia CI looks like.  Your build script then downloads a prebuilt julia [3], and invokes Revise on it.  Even though the folder is named `julia-buildkite`, it should be the `julia` repository.


[0]: https://buildkite.com/julialang/julia-buildkite/builds/1646#01910d9b-e1b1-4e5b-a1af-bf4ffea85858/117-123
[1]: https://buildkite.com/julialang/julia-buildkite/builds/1646#01910d9b-e1b1-4e5b-a1af-bf4ffea85858/165-167
[2]: https://buildkite.com/julialang/julia-buildkite/builds/1646#01910d9b-e1b1-4e5b-a1af-bf4ffea85858/197-199
[3]: https://buildkite.com/julialang/julia-buildkite/builds/1646#01910d9b-e1b1-4e5b-a1af-bf4ffea85858/609-611

**Comment by Keno** [comment]
What do we need to do to push this forward. Can we just rm the julia source dir for now?

**Comment by KristofferC** [comment]
The Revise step should pass now and this should be good to go (but let's confirm).

**Review by IanButterworth — APPROVED** [review]

**Comment by KristofferC** [comment]
The gnuassertrr and windows failure are not related to this PR I guess? But they are required for merging?

**Comment by DilumAluthge** [comment]
I'll retry all the jobs. Let's see how we do the second time around.

**Comment by DilumAluthge** [comment]
Yeah, we have all the jobs marked as required, because merging a broken PR in this repo can instantly break all of base CI, so we figured we'd lean towards the more conservative side.

**Comment by DilumAluthge** [comment]
Alright, CI is all green now.

@KristofferC and @IanButterworth: is this PR good to merge?

**Comment by DilumAluthge** [comment]
I see

```
Testing Revise tests passed
```

in the logs, which presumably is a good sign.

**Pull request merged**